File: GriffBallJava/GriffServer/src/ClientHandler.java
import java.io.PrintWriter;
import java.net.Socket;
import java.util.Scanner;

public class ClientHandler implements Runnable {
    private final Socket socket;
    private GriffBall griff;
    public static int idGenerator = 1;

    public ClientHandler(Socket socket, GriffBall griff) {
        this.socket = socket;
        this.griff = griff;
    }

    @Override
    public void run() {
        int playerId = idGenerator++;
        try (Scanner reader = new Scanner(socket.getInputStream());
             PrintWriter writer = new PrintWriter(socket.getOutputStream(), true)) {
            try {
                String playerName = reader.nextLine();
                griff.addPlayer(playerId, playerName);
                System.out.println("Player_" + playerId + "(" + playerName + ") has joined the game");
                griff.printPlayers();
                writer.println("success " + playerId);

                while (true) {
                    String line = reader.nextLine();
                    String[] substrings = line.split(" ");
                    griff.checkCorrectState();
                    switch (substrings[0].toLowerCase()) {
                        case "ball":
                            String playerHasBall = griff.whoHasBall().trim();
                            writer.println(playerHasBall);
                            break;
                        case "throw":
                            int toPlayerId = Integer.parseInt(substrings[1]);
                            griff.giveBall(toPlayerId);
                            break;
                        case "players":
                            String playerString = griff.getPlayersString();
                            writer.println(playerString);
                            break;
                        default:
                            throw new Exception("Unknown command: " + substrings[0]);

                    }
                    // griff.checkCorrectState();
                }

            } catch (Exception e) {
                writer.println("ERROR" + e.getMessage());
            }
        } catch (Exception e) {
        } finally {
            System.out.println("Player " + playerId + " has left the game");
            griff.removePlayer(playerId);
            griff.printPlayers();
        }
    }
}

File: GriffBallServer/ClientHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;

namespace GriffBallServer
{
    class ClientHandler
    {
        private TcpClient tcpClient;
        private GriffBall griff;
        public static int idGenerator = 1;

        public ClientHandler(TcpClient tcpClient, GriffBall griff)
        {
            this.tcpClient = tcpClient;
            this.griff = griff;
            this.run();
        }

        public void run()
        {
            using (NetworkStream stream = tcpClient.GetStream())
            {
                StreamWriter writer = new StreamWriter(stream);
                StreamReader reader = new StreamReader(stream);
                int playerId = idGenerator++;
                try
                {
                    string playerName = reader.ReadLine();
                    griff.addPlayer(playerId, playerName);                    
                    Console.WriteLine("Player_" + playerId + "(" + playerName + ") has joined the game");
                    griff.printPlayers();
                    writer.WriteLine("success " + playerId);
                    writer.Flush();

                    while (true)
                    {
                        string line = reader.ReadLine();
                        string[] substrings = line.Split(' ');
                        griff.checkCorrectState();
                        switch (substrings[0].ToLower())
                        {
                            case "ball":
                                string playerHasBall = griff.whoHasBall().Trim();
                                writer.WriteLine(playerHasBall);
                                writer.Flush();
                                break;

                            case "throw":
                                int toPlayerId = Convert.ToInt32(substrings[1]);
                                griff.giveBall(toPlayerId);
                                break;

                            case "players":
                                string playerString = griff.getPlayersString();
                                writer.WriteLine(playerString);
                                writer.Flush();
                                break;

                            default:
                                throw new Exception($"Unknown command: {substrings[0]}.");
                        }
                    }
                }
                catch (Exception e)
                {
                    writer.WriteLine("ERROR" + e.Message);
                }
                finally
                {
                    try
                    {
                        griff.removePlayer(playerId);
                        writer.Flush();
                        tcpClient.Close();
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine("ERROR CLOSING:" + e.Message);
                    }

                    Console.WriteLine($"Player_{playerId} has left the game");
                    griff.printPlayers();
                }
            }
        }
    }
}
